File: src/backend/exe/network/transport/ipc/IpcTransport.cpp
#include "IpcTransport.hpp"

#include <sys/socket.h>
#include <unistd.h>

#include <cstring>
#include <iostream>

#include "Logger.hpp"

/**
 * @brief Constructs a POSIX Unix domain socket transport with an existing socket.
 * @param socket_fd The file descriptor of the connected Unix socket.
 */
IpcTransport::IpcTransport(int socket_fd) : fd(socket_fd) {}

/**
 * @brief Destructor ensures that the socket is closed.
 */
IpcTransport::~IpcTransport() {
    close();
}

/**
 * @brief Sets callback to be invoked when connection closes unexpectedly.
 * @param onClose Callback function to handle connection closure.
 */
void IpcTransport::setCloseCallback(CloseCallback onClose) {
    closeCallback_ = onClose;
}

/**
 * @brief Starts receiving data on the transport.
 * @param onReceive Callback function to handle received data.
 */
void IpcTransport::start(ReceiveCallback onReceive) {
    // Required since nothing prevents start() from being called twice for the same instance.
    if (running.exchange(true))
        return;

    auto& logger = Logger::instance();
    logger.trace("Starting reader thread for Unix socket fd " + std::to_string(fd));

    readerThread = std::jthread([this, onReceive](std::stop_token st) {
        auto& logger = Logger::instance();
        logger.trace("Reader thread started for Unix socket fd " + std::to_string(fd));

        char buffer[1024];
        bool connection_closed_by_peer = false;

        while (!st.stop_requested() && running.load()) {
            logger.trace("Calling read() on Unix socket fd " + std::to_string(fd));
            ssize_t n = read(fd, buffer, sizeof(buffer));
            logger.trace("read() returned " + std::to_string(n) + " for Unix socket fd " + 
                        std::to_string(fd));

            if (n <= 0) {
                if (n == 0) {
                    logger.trace("Client disconnected (EOF) on Unix socket fd " + 
                                std::to_string(fd));
                    connection_closed_by_peer = true;
                } else {
                    logger.error("Read error on Unix socket fd " + std::to_string(fd) + ": " +
                                 std::string(strerror(errno)));
                    connection_closed_by_peer = true;
                }
                running = false;
                break;
            }

            std::string payload(buffer, n);
            onReceive(payload);
        }

        logger.trace("Reader thread EXITING for Unix socket fd " + std::to_string(fd));

        // Notify session that connection died
        if (connection_closed_by_peer && closeCallback_) {
            logger.trace("Invoking close callback for Unix socket fd " + std::to_string(fd));
            closeCallback_();
        }
    });
}

/**
 * @brief Sends data over the transport.
 * @param data The data to send.
 */
void IpcTransport::send(const std::string& data) {
    if (!running.load()) {
        return;
    }

    ssize_t sent = write(fd, data.data(), data.size());
    if (sent < 0) {
        auto& logger = Logger::instance();
        logger.error("Write error on Unix socket fd " + std::to_string(fd) + ": " +
                     std::string(strerror(errno)));
        running = false;
    }
}

bool IpcTransport::connect() {
    // The socket is already connected, so just return true.
    return true;
}

/**
 * @brief Closes the Unix socket and terminates the reading loop.
 */
void IpcTransport::close() {
    if (!running.exchange(false))
        return;

    auto& logger = Logger::instance();
    logger.debug("Closing Unix socket transport on fd " + std::to_string(fd));

    // Close socket first to unblock read()
    if (fd >= 0) {
        shutdown(fd, SHUT_RDWR);
        ::close(fd);
        fd = -1;
    }

    logger.debug("Unix socket transport closed");
}

File: src/frontend/network/transport/ipc_transport.py
import os
import socket
import threading
from typing import Callable, Optional

from network.transport.transport_interface import ITransport
from utils.logger import Logger


class IpcTransport(ITransport):
    """Class for Unix domain socket (IPC) transport"""

    def __init__(self, fd: int):
        """
        Construct an IPC transport with an existing socket file descriptor.

        @param fd The file descriptor of the connected Unix socket.
        """
        self.fd = fd
        self.running = False
        self.reader_thread: Optional[threading.Thread] = None
        self.logger_ = Logger()

    def __del__(self):
        """Destructor ensures that the socket is closed."""
        self.close()

    def start(self, on_receive: Callable[[str], None]) -> None:
        """
        Start receiving data on the transport.

        @param on_receive Callback function to handle received data.
        """
        if self.fd < 0:
            self.logger_.error("Cannot start: invalid file descriptor")
            return

        self.running = True
        self.logger_.debug(f"Starting reader thread for Unix socket fd {self.fd}")

        def reader_loop():
            self.logger_.debug(f"Reader thread started for Unix socket fd {self.fd}")

            while self.running:
                try:
                    # Read from socket (max 4096 bytes)
                    data = os.read(self.fd, 4096)

                    # Connection closed or broken
                    if len(data) == 0:
                        self.logger_.info("Server closed connection")
                        self.running = False
                        break

                    payload = data.decode("utf-8")
                    on_receive(payload)

                except OSError as e:
                    if self.running:
                        self.logger_.error(
                            f"Read error on Unix socket fd {self.fd}: {e}"
                        )
                    self.running = False
                    break
                except Exception as e:
                    self.logger_.error(
                        f"Unexpected read error on Unix socket fd {self.fd}: {e}"
                    )
                    self.running = False
                    break

            self.logger_.debug(f"Reader thread EXITING for Unix socket fd {self.fd}")

        self.reader_thread = threading.Thread(target=reader_loop, daemon=True)
        self.reader_thread.start()

    def send(self, data: str) -> None:
        """
        Send data over the transport.

        @param data The data to send.
        """
        if self.fd < 0:
            self.logger_.error("Cannot send: socket closed")
            return

        if not self.running:
            return

        try:
            os.write(self.fd, data.encode("utf-8"))
        except OSError as e:
            self.logger_.error(f"Write error on Unix socket fd {self.fd}: {e}")
            self.running = False
        except Exception as e:
            self.logger_.error(
                f"Unexpected send error on Unix socket fd {self.fd}: {e}"
            )
            self.running = False

    def close(self) -> None:
        """Close the Unix socket and terminate the reading loop."""
        if not self.running:
            return

        self.running = False
        self.logger_.debug(f"Closing Unix socket transport on fd {self.fd}")

        if self.fd >= 0:
            try:
                os.close(self.fd)
            except Exception:
                pass
            finally:
                self.fd = -1

        # Wait for reader thread to finish
        if self.reader_thread and self.reader_thread.is_alive():
            self.reader_thread.join(timeout=1.0)

        self.logger_.debug("Unix socket transport closed")
